src/ngraph/placement.cpp

```diff
@@ -224,3 +224,186 @@ pair<vector<shared_ptr<Function>>, unordered_map<shared_ptr<op::Parameter>, shar
+        sorted_nodes.push_back(node_map.at(independent_node));
+
+        for (auto user : independent_node->get_users())
+        {
+            Node* user_node = user.get();
+            node_dependency_count.at(user_node) -= 1;
+            if (node_dependency_count.at(user_node) == 0)
+            {
+                independent_nodes_by_placement[user_node->get_placement_size()].push_back(
+                    user_node);
+            }
+        }
+    }
+
+    if (sorted_nodes.size() != f->get_ops().size())
+    {
+        throw ngraph_error("sorted_nodes.size()== " + to_string(sorted_nodes.size()) +
+                           " != f->get_ops().size()== " + to_string(f->get_ops().size()) +
+                           ". Internal error with topological sort.");
+    }
+
+    // Build clusters from the sorted_nodes
+    previous_placement = 0; // Placement::DEFAULT;
+    vector<unordered_set<shared_ptr<Node>>> clusters;
+    for (shared_ptr<Node> node : sorted_nodes)
+    {
+        size_t node_placement = node->get_placement_size();
+        if (node_placement != previous_placement)
+        {
+            unordered_set<shared_ptr<Node>> new_cluster;
+            clusters.push_back(new_cluster);
+        }
+        clusters.back().insert(node);
+        previous_placement = node_placement;
+    }
+
+    // Sanity check for node duplication and full node coverage
+    unordered_set<shared_ptr<Node>> cluster_nodes;
+    for (auto cluster : clusters)
+    {
+        for (auto node : cluster)
+        {
+            if (cluster_nodes.find(node) != cluster_nodes.end())
+            {
+                throw ngraph_error("Node " + node->get_name() + " is duplicated in clusters");
+            }
+            cluster_nodes.insert(node);
+        }
+    }
+    unordered_set<shared_ptr<Node>> f_nodes;
+    for (auto node : f->get_ordered_ops())
+    {
+        f_nodes.insert(node);
+    }
+    if (cluster_nodes != f_nodes)
+    {
+        throw ngraph_error(
+            "Cluster's nodes are not the same as function's nodes. cluster_nodes.size()=" +
+            to_string(cluster_nodes.size()) + ", f_nodes.size()=" + to_string(f_nodes.size()));
+    }
+
+    return clusters;
+}
+
+// Suffix *_size  as a part of function name is temporary, this suffix
+//  will be removed when the backends move to the latest Hybrid backend
+pair<vector<shared_ptr<Function>>, unordered_map<shared_ptr<op::Parameter>, shared_ptr<op::Result>>>
+    ngraph::split_function_by_placement_size(const shared_ptr<Function>& f)
+{
+    // Split functions to clusters of nodes that can be computed together
+    vector<unordered_set<shared_ptr<Node>>> clusters = group_function_nodes_to_clusters_size(f);
+
+    // Map from (intermediate) parameter to result node, for guiding data copy among devices
+    unordered_map<shared_ptr<op::Parameter>, shared_ptr<op::Result>> map_parameter_to_result;
+
+    // Split neighboring nodes if they belong to different clusters
```

Comment:
nitpick:
can we please add another check that GOEs end up in the same cluster as their multi-output op?